### PR TITLE
Xenomorphs cannot use most abilities nor attack while on a solid turf (e.g. a resin wall)

### PR DESCRIPTION
**Pull request by DeltaFire15**
## About The Pull Request
This is intended as an alternate PR to the drone / general building changes in #9496.

While builders building a wall below themselves and then doing things like aciding walls or simillar is annoying, I believe removing building below oneself alltogether is too much of an impedement to builders in general. Due to the fact that the only real issue from allowing this is castes being able to use walls as a "blocker" for shots / turrets while using abilities from the same turf, I simply disabled using abilities or attacking while being on such a turf.
~~Two~~ A few exceptions to this apply:
Attacking walls, as resin walls use this to tear them down (and xenos cannot do anything bad to walls simply via base attack), is still allowed, so you can tear down walls if you got yourself stuck (oops!)
The resting toggle can still be used on a closed turf because it'd be kinda weird if you couldn't.
Edit:
Hivemind's Change Form and Wraith's Phase-Shift abilities can also still be used while in walls. No other ones though.

Note, this does not apply any changes to hivemind's building like the other PR did in parallel. Just tell me if you want me to do something simillar in this one, or just atomize that particular change.

(also fixes XACT_DO_AFTER_ATTACK not being in the bitfield lists in bitfield.dm, but thats a one line change)
Edit:
~~Additionally, fixes a turret bug that made them behave weirdly at the edge of their range and another that made them runtime when their target left their range.~~ atomized.
To prevent a cheese replacement, resin doors now start open when built on your own tile (but will close after a bit once you leave the tile, no need to manually close every door either)

## Why It's Good For The Game
Removing combat walling to melt / spit things with little resistance from especially turrets is pretty reasonable, but inconveniencing builders in general seems overboard. This applies a (in my opinion) reasonable middle-ground that should satisfy both parties, as using abilities while on solid turfs shouldn't really be the case and only applies some funky cheese vectors.

## Changelog
:cl:
balance: Xenomorphs cannot use abilities nor attack while on a solid turf (e.g. a wall). This excludes attacking walls so you can still break your resin wall if you got yourself stuck aswell as the "toggle resting" ability and Hivemind's change form & Wraith's phase-shift / resync.
tweak: Resin doors start open if built on your own tile.
fix: Added XACT_DO_AFTER_ATTACK to bitfields.dm as it was missing (aka it gets "translated" into readable language during varediting of that var)
/:cl:

### CODE REVIEW

**Comment by vilereaver41** [comment]
Agreed what a cheapo tactic

**Comment by Luxyboo** [comment]
does this mess with xenos on top of crushers?

**Comment by DeltaFire15** [comment]
> does this mess with xenos on top of crushers?

Crushers are not a turf, so unless said crusher is in a wall, no.

File: code/_onclick/xeno.dm [hunk 1]
```diff
@@ -1,7 +1,10 @@
 /mob/living/carbon/xenomorph/UnarmedAttack(atom/A, has_proximity, modifiers)
 	if(lying_angle)
 		return FALSE
-
+	var/turf/current_turf = get_turf(src)
+	if(isclosedturf(current_turf) && !iswallturf(A))	//If we are on a closed turf (e.g. in a wall) we can't attack anything, except walls (or well, resin walls really) so we can't make ourselves be stuck.
```
Comment (BraveMole) [inline code/_onclick/xeno.dm:5]:
```suggestion
	if(isclosedturf(get_turf(src)) && !iswallturf(A))	//If we are on a closed turf (e.g. in a wall) we can't attack anything, except walls (or well, resin walls really) so we can't make ourselves be stuck.
```

**Review by BraveMole — CHANGES_REQUESTED** [review]
I much prefer this but it was tivi who didn't want it. Implementation is good though

**Comment by Frampox** [comment]
Looks nice. What about mind and wrath?

**Comment by DeltaFire15** [comment]
> Looks nice. What about mind and wrath?



> Looks nice. What about mind and wrath?

Hivemind: 
Should disable doing stuff on walls, which is simillar to cheeky moves of other xenos. Teleporting from a wall should in theory still work since it's not a normal ability but I'll check.
I should however make the demanifest ability able to be used while on one.

Wraith:
Cannot do teleport cheese via teleporting into a wall, grabbing something on one side and teleporting them to the other anymore (especially 'fun' when you can brazil marines to the other side of a wall in e.g. SD if they don't move immediately because they were fixing a cade). I should however allow using the ability that allows you to return to the starting point of phasing, whatever it was named.

**Comment by DeltaFire15** [comment]
Requested changes aswell as the things I outlined in the previous comment have been applied. Additionally, to avoid an easy cheese replacement, resin doors start open if constructed on your own tile (but they'll obviously just autoclose a moment after you leave the tile). Yes, doors may be terrible durability-wise, but that doesn't matter if you don't get targetted at all due to turret vision AI.
I also found two turret bugs while testing door cheese, so I fixed those aswell.

**Comment by Lumipharon** [comment]
This also fixes xenos being able to vent crawl out onto walls (mainly on crash) and cheesing stuff.

**Comment by DeltaFire15** [comment]
> This also fixes xenos being able to vent crawl out onto walls (mainly on crash) and cheesing stuff.

.. I didn't even consider beans just building walls over vents and then ventcrawling on top of said vent. Buut, yeah this should avoid that too (you can still exit on top of a wall, but no doing funky stuff at least)

**Comment by Lumipharon** [comment]
> .. I didn't even consider beans just building walls over vents and then ventcrawling on top of said vent. Buut, yeah this should avoid cheese related to that too (you can still exit on top of a wall, but no doing funky stuff at least)

Nah, what happens is the Ship on crash destroys everything under it when it lands, so if it deletes some pipes, a vent crawling beno can just exit the broken pipe directly onto the outerhull turf, and spray acid/melt shit while being completely safe, then just moon walk out when they're done.

File: code/modules/projectiles/sentries.dm [hunk 2]
```diff
@@ -373,7 +373,10 @@
 	var/obj/item/weapon/gun/gun = internal_item
 	var/mob/living/target = get_target()
 	update_icon()
-	if(target != gun.target || get_dist(src, target) > range)
+	if(!target || get_dist(src, target) > range)
+		gun.stop_fire()
+		return
+	if(target != gun.target)
```
[the remark this answers is not in the document]
Comment (DeltaFire15) [reply code/modules/projectiles/sentries.dm:379]: No impact on balance save for fixing incorrect or runtiming behavior, but sure.

**Comment by Lumipharon** [comment]
> @Lumipharon nope, vents no longer get deleted when tadpole/caunter lands ontop of them, this includes vent entrances so xenos can (currently) ventcrawl directly into caunter if it landed on a vent

Ah, so that's what's happening. This PR fixes it either way though.

**Pull request merged**